### ISSUE TITLE
Issue with test_model.py not yielding satisfactory results after training with train.py

### ISSUE THREAD

**Issue by T-matsubara** [ibaiGorordo/CREStereo-Pytorch#19]
Dear GitHub Community,

I hope this message finds you well. I am reaching out to seek clarification regarding the training duration for my project, as well as any possible remedies for the unsatisfactory results obtained.

Firstly, I would like to address the training duration. During the training process, it took approximately 10 days to complete. However, I wanted to confirm if this duration is within the expected range or if it seems unusually long. Considering the following training conditions:

nr_gpus: 2
batch_size: 6
n_total_epoch: 600
minibatch_per_epoch: 500
I would greatly appreciate your insights or any available information to determine if the 10-day training duration aligns with expectations or if there might be any potential issues causing the extended duration.

Secondly, I encountered unsatisfactory results when executing the test_model.py script after training with train.py. Despite following the instructions and procedures correctly, the outcome did not meet my expectations. I utilized test images downloaded from the CREStereo GitHub repository and ensured that the images were sourced exclusively from this repository. To help diagnose the issue more effectively, I have attached a photo illustrating the execution results for your reference.

I would like to kindly request any guidance, suggestions, or potential remedies you can provide to address the unsatisfactory results. Your expertise in identifying any potential issues and proposing solutions would be immensely valuable.

Thank you for your attention to this matter. I eagerly await your response.
![output_720](https://github.com/ibaiGorordo/CREStereo-Pytorch/assets/128988709/2fccc060-dd4d-4c15-ac20-43e8dc0401e2)

**Comment by dadaokanxigua** [comment]
Hi @T-matsubara,
Firstly, You can set the "num_workers" in the "DataLoader" more than 0 to reduce the training time.

Secondly, I also failed to train my model which performace is as good as [Megengine Weights](https://drive.google.com/file/d/1Wx_-zDQh7BUFBmN9im_26DFpnf3AkXj4/view), but I think it is close.
Here is my result:
![image](https://github.com/ibaiGorordo/CREStereo-Pytorch/assets/23627602/4d1a0a77-53a8-418b-9a6f-d087bfafd231)

My training conditions:
nr_gpus: 1
batch_size: 16
n_total_epoch: 600
minibatch_per_epoch: 500/16
The reason of number of nr_gpus and minibatch is lack of GPU resouces. I believe I will get better result if enlarge the mini_batch_per_epoch, and I will try it later.

You can also try this training condition. I look forward to discussing your new results with you.

**Comment by T-matsubara** [comment]
Thank you for your response! I greatly appreciate your assistance, as it helped me confirm that I can successfully train using train.py. I have observed that using a batch size of 16 puts a significant load on the GPU. Therefore, I would like to inquire about the specifications of the GPU you used for execution and the time it took for the training to complete. Additionally, I am curious about the reasoning behind setting the value of minibatch_per_epoch to 500/16.

Thank you once again for your support.

Best regards.

**Comment by T-matsubara** [comment]
@dadaokanxigua 
Thank you for your response! I greatly appreciate your assistance, as it helped me confirm that I can successfully train using train.py. I have observed that using a batch size of 16 puts a significant load on the GPU. Therefore, I would like to inquire about the specifications of the GPU you used for execution and the time it took for the training to complete. Additionally, I am curious about the reasoning behind setting the value of minibatch_per_epoch to 500/16.

Thank you once again for your support.

Best regards.

**Comment by T-matsubara** [comment]
Dear @dadaokanxigua 
Thank you for providing me with the detailed experimental conditions. I sincerely appreciate your help in sharing that information. As soon as my training is completed, I will make an effort to replicate your conditions and conduct the training again. Your kind and thorough assistance is highly valued, and I am truly grateful for it.

Once again, thank you for your support.

Best regards.

**Comment by T-matsubara** [comment]
Dear @dadaokanxigua 
Thank you for providing me with the detailed experimental conditions. I sincerely appreciate your help in sharing that information. As soon as my training is completed, I will make an effort to replicate your conditions and conduct the training again. The additional conditions you have mentioned, including nr_gpus: 2, batch_size: 2, n_total_epoch: 600, minibatch_per_epoch: 500*8, and "num_workers" = 2, will be taken into account during my next attempt.

Your kind and thorough assistance is highly valued, and I am truly grateful for it.

Once again, thank you for your support.

**Comment by T-matsubara** [comment]
Dear @dadaokanxigua 

I executed the code with the following conditions: 
nr_gpus: 2, 
batch_size: 2, 
n_total_epoch: 600, 
minibatch_per_epoch: 500*8, 
line 355 with num_workers=2, drop_last=True, persistent_workers=False, pin_memory=True. However, the execution time has become 480 hours. I believe the possible cause could be increasing the minibatch_per_epoch. Based on these circumstances, I have a few questions:

First, are you using DDP (Distributed Data Parallel) in your code?

Second, as my knowledge and understanding might be insufficient, I would appreciate it if you could point out any mistakes I might have made.

Thank you in advance.

**Comment by dadaokanxigua** [comment]
Hi @T-matsubara ,
Firstly, The reason for the increase in training time is indeed because of the increase in minibatch_per_eopch.
And I used the DDP function.

Your batchsize does not need to be set so small, I set it to 2 because my GPU memory is not enough to set more than 2.
In fact, larger batchsize setting is better, so that your minibatch_per_epoch can be set smaller, thereby reducing training time.
Moreover, larger batches are easier to converge.

It seems to me that your network is able to converge under the initial training conditions. Can you confirm that you used multiple cards for training? If you run the code of line 355, you should be in single-card training mode.

**Comment by T-matsubara** [comment]
Dear @dadaokanxigua 
Thank you for your continued engagement and the additional information provided. I understand that you have rewritten the code and executed it, and upon checking with nvidia-smi, you found that the PIDs were different, leading you to believe that DDP was not being executed.

I appreciate your efforts in troubleshooting the issue on your own. It's important to ensure that all necessary steps are taken to enable successful DDP execution. In light of this, I would be grateful if you could kindly share the approach or modifications you made to the code that allowed you to execute DDP successfully. Understanding your solution will not only help in resolving the issue but also broaden our knowledge base for future reference.

I look forward to your response and learning more about the modifications you made to achieve successful DDP execution. Thank you for your cooperation.

I apologize for repeatedly asking questions.

Here is the code I executed.
`CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch --nproc_per_node=2 train.py`

Best regards.

**Comment by T-matsubara** [comment]
I believe the error you encountered was likely caused by inserting the latest version of PyTorch when creating a Docker image. As a result, you modified the code on line 90 to change the DDP format to torchrun. Below, I've included the modified train_dist code and the execution statement.

`def train_dist(args, world_size):
    local_rank = int(os.environ['LOCAL_RANK'])
    # directory check
    log_model_dir = os.path.join(args.log_dir, "models")
    ensure_dir(log_model_dir)`
    
    `CREStereo-Pytorch# CUDA_VISIBLE_DEVICES=0,1 torchrun --nproc_per_node=2 train.py
WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
2023/07/13 21:34:54 Use 2 GPU(s)
2023/07/13 21:34:54 Params: 5432948
2023/07/13 21:34:56 Dataset size: 200000
INFO:torch.nn.parallel.distributed:Reducer buckets have been rebuilt in this iteration.`

While training, you have observed that the PIDs displayed by nvidia-smi are different, indicating an increase in computational speed. Additionally, you have noticed that the loss is converging. Although it's uncertain if you'll obtain the correct results, you expect to know the outcome in two days. Thank you for your collaboration so far.
